### PR TITLE
Customize pass group ordering

### LINKED ISSUE #562: Allow for customized ordering of passes
Allow the user to specify any order in which the passes are run.

### PR DESCRIPTION
**Pull request by jdahm**
## Technical Description

Allow pass groups to be run in a custom order.

## Dependencies

Resolves #562.

### CODE REVIEW

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
I worked with the existing framework when implementing this. It's not ideal, but it does allow custom groups to be run through `DawnCompiler::optimize()` and `DawnCompiler::compile()` methods by adding a second argument that defaults to an empty list.

File: dawn/src/dawn/Compiler/DawnCompiler.cpp [hunk 1]
```diff
@@ -481,7 +445,7 @@ DawnCompiler::compile(const std::shared_ptr<SIR>& stencilIR) {
   }
 
   // Optimize the IIR
-  auto optimizedStencilInstantiation = optimize(stencilInstantiation);
+  auto optimizedStencilInstantiation = optimize(stencilInstantiation, groups);
```
Comment (eddie-c-davis) [inline dawn/src/dawn/Compiler/DawnCompiler.cpp]: Nice refactoring!

File: dawn/src/dawn/Compiler/DawnCompiler.h [hunk 2]
```diff
@@ -20,12 +20,31 @@
...
 #include "dawn/Support/NonCopyable.h"
+#include <list>
 #include <memory>
 
 namespace dawn {
 
 struct SIR;
 
+/// @brief Enumeration of all pass groups
+enum class PassGroup {
```
Comment (eddie-c-davis) [inline dawn/src/dawn/Compiler/DawnCompiler.h]: Like the enumeration.

File: dawn/src/dawn/Compiler/DawnCompiler.cpp [hunk 3]
```diff
@@ -250,138 +228,124 @@ DawnCompiler::optimize(std::map<std::string, std::shared_ptr<iir::StencilInstant
...
-    optimizer.pushBackPass<PassInlining>(
-        (getOptions().Inlining || getOptions().TmpToStencilFunction),
-        PassInlining::InlineStrategy::ComputationsOnTheFly);
-    // validation check
-    optimizer.pushBackPass<PassValidation>();
-  }
-  //===-----------------------------------------------------------------------------------------
-  if(shouldRunGroup(options_, options_.IntervalPartitioning)) {
-    if(options_.IntervalPartitioning) {
+  if(groups.empty())
```
Comment (twicki) [inline dawn/src/dawn/Compiler/DawnCompiler.cpp]: shouldn't this be the responsibility of the caller of optimize? Feels out of place here

File: dawn/src/dawn/Compiler/DawnCompiler.cpp [hunk 4]
```diff
@@ -250,138 +228,124 @@ DawnCompiler::optimize(std::map<std::string, std::shared_ptr<iir::StencilInstant
...
+    groups = defaultPassGroups();
+
+  for(auto g : groups) {
```
Comment (twicki) [inline dawn/src/dawn/Compiler/DawnCompiler.cpp]: g is not an indicative name

File: dawn/src/dawn/Compiler/DawnCompiler.h [hunk 5]
```diff
@@ -20,12 +20,31 @@
...
+  Parallel,
```
Comment (twicki) [inline dawn/src/dawn/Compiler/DawnCompiler.h]: what is the order of these?

File: dawn/src/dawn/Optimizer/Options.inc [hunk 6]
```diff
@@ -47,6 +47,7 @@ OPT(int, BlockSizeI, 0, "block-size-i", "", "i block size for tiled computations
 OPT(int, BlockSizeJ, 0, "block-size-j", "", "j block size for tiled computations", "", true, false)
 OPT(int, BlockSizeK, 0, "block-size-k", "", "k block size for tiled computations", "", true, false)
 
+// TODO Remove DefaultNone
```
Comment (twicki) [inline dawn/src/dawn/Optimizer/Options.inc]: this should not be merged

**Comment by jdahm** [comment]
launch jenkins

File: dawn/src/dawn/Compiler/DawnCompiler.cpp (hunk 3 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/DawnCompiler.cpp]: Will be done in a follow-up PR.

File: dawn/src/dawn/Compiler/DawnCompiler.h (hunk 5 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/DawnCompiler.h]: The ordering isn't defined by the enumeration, although that is the order they were initially called in. See `defaultPassGroups()`.

File: dawn/src/dawn/Compiler/DawnCompiler.h (hunk 2 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/DawnCompiler.h]: That came from your enumeration in `CompilerUtil`. Good work! 👍

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Pull request merged**